Fix misleading log message for aria-current="true"

feerrenrut opened the pull request:
### Link to issue number:
None

### Summary of the issue:
There is a misleading log message for aria-current="true". It was a non-handled value, an exception was raised, a message added to the log, and the value `True` (bool rather than str) was used.

When trying to follow this code it is hard to know where the values are strings, None, or bool. In some cases the naming is overly specific.

### Description of how this pull request fixes the issue:

Types and naming for isCurrent were confusing, so normalize the types
using an enum and only name as ariaCurrent when the data is specific to aria-current,
otherwise use 'isCurrent'. Include comments to link aria-current as the
potential (likely) source for 'isCurrent'

### Testing performed:
Used Chrome, Firefox, IE, and Edge with https://a11ysupport.io/tests/html/aria/aria-current.html

### Known issues with pull request:
This is technically an API breaking change. If there are any concerns about this, we may have to wait for the 2021.1 release. 

### Change log entry:

```
changes for developers:
- 'NVDAObject' (and derivatives) property 'isCurrent' now strictly returns 'controlTypes.IsCurrent'. It is no longer Optional, and thus will not return None. When an object is not current 'controlTypes.IsCurrent.NO' is returned.
- The 'controlTypes.isCurrentLabels' has been removed, instead use the 'displayString' property on a 'controlTypes.IsCurrent' enum value. EG 'controlTypes.IsCurrent.YES.displayString'
```

feerrenrut committed:
Fix misleading log message for aria-current="true"

Types and naming for isCurrent were confusing, so normalize the types
using an enum and only name as ariaCurrent when the data is specific to aria-current,
otherwise use 'isCurrent'. Include comments to link aria-current as the
potential (likely) source for 'isCurrent'

feerrenrut committed:
Add type information for auto property

feerrenrut committed:
fix lint

leonardder commented on source/controlTypes.py:
```diff
@@ -662,21 +664,46 @@ class OutputReason(Enum):
+	DATE = "date"
+	TIME = "time"
+
+	def getTranslationForUI(self):
```
Just a thought, if you'd make this:
```suggestion
	def __str__(self):
```

Then, you could do str(value), which is a lot shorter and I think if done consistently (e.g. for roles and states in the future), it is more straight forward.
May be you find the current method more obvious, though.

feerrenrut replied:
I think this might cause some confusion, I want it to be clear that the returned string is translated, and not just a `str` representation of the enum.

leonardder commented on source/controlTypes.py:
```diff
@@ -662,21 +664,46 @@ class OutputReason(Enum):
-#: within a particular kind of selection.
-isCurrentLabels: Dict[Union[bool, str], str] = {
+#: within a particular kind of selection. EG aria-current
+isCurrentLabels: Dict[IS_CURRENT, str] = {
```
Could this be moved onto the enum?, as it is now strictly private to the enum?

feerrenrut replied:
Yes, I think so.

feerrenrut replied:
This was a bit of a rabbit hole. Turns out moving `isCurrentLabels` into the class causes it to be considered an Enum Member. This ends up being a bit confusing. I think it is better outside of the class, but I have prefixed an underscore to mark it as module private.

leonardder replied:
Ah, sounds ok to me.

feerrenrut commented on source/speech/__init__.py:
```diff
@@ -1606,15 +1606,12 @@ def getPropertiesSpeech(  # noqa: C901
+	# speak isCurrent property EG aria-current
+	isCurrent = propertyValues.get('current', controlTypes.IS_CURRENT.NO)
+	if isCurrent != controlTypes.IS_CURRENT.NO:
+		textList.append(controlTypes.IS_CURRENT.getTranslationForUI(isCurrent))
```
I think this could be 
```suggestion
		textList.append(isCurrent.getTranslationForUI())
```

feerrenrut committed:
Make isCurrent enum to translation private

feerrenrut committed:
Shorten calls to getTranslationForUI

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into fixMisleadingAriaCurrentLogWarning

feerrenrut committed:
_isCurrentLabels was being converted into an Enum Member

leonardder commented on source/controlTypes.py:
```diff
@@ -662,21 +664,46 @@ class OutputReason(Enum):
+	DATE = "date"
+	TIME = "time"
+
+	def getTranslationForUI(self):
```
Alternative idea

```suggestion
	@property
	def translationForUI(self):
```

I think translation or translated would also be sufficient.

feerrenrut committed:
review actions

- Use a property
- shorten name

feerrenrut commented:
@leonardder I have addressed your suggestions. I think making this a property is a good idea. I decided against `translation` because it is hard to search for usages in the code base. Instead I have gone with `displayString` which communicates that it is for displaying to the user. This also leaves room for other transformations (other than translation) to happen to this string prior to being presented to a user.

leonardder commented on source/controlTypes.py:
```diff
@@ -662,21 +664,50 @@ class OutputReason(Enum):
-#: within a particular kind of selection.
-isCurrentLabels: Dict[Union[bool, str], str] = {
+
+class IS_CURRENT(Enum):
```
I must have missed this. I think it makes sense to follow Python conventions here, I believe naming should be:
```suggestion
class IsCurrent(Enum):
```

Upper case enum values are ok

feerrenrut committed:
review: Use pascal case for IsCurrent

LeonarddeR reviewed (dismissed):
Great @feerrenrut

feerrenrut commented:
I have just noticed inconsistent case for `IsCurrent.Yes` all the rest of the values are uppercase. I'll fix this.

feerrenrut commented:
[Python docs for enum](https://docs.python.org/3/library/enum.html) use upper case for members of the enum, so I'll match that style.

feerrenrut committed:
Upper case for enum members

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into fixMisleadingAriaCurrentLogWarning

feerrenrut committed:
update changes file for PR #11782

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into fixMisleadingAriaCurrentLogWarning

feerrenrut committed:
fix lint

feerrenrut committed:
Update change log for PR #11782
